gazebo_ros/src/node.cpp

```diff
@@ -38,7 +38,7 @@ Node::SharedPtr Node::Get(sdf::ElementPtr sdf)
 {
   // Initialize arguments
   std::string name = "";
-  std::string ns = "";
+  std::string ns = "/";
```

Comment:
Valid namespaces should be always absolute (contrary to names, which can be relative/private/absolute).

I guess @jacobperron did that change because the namespace is being passed to the new QoS class, and some of the API it uses may be doing namespace validation.

@jacobperron can you confirm if that was the case?

Reply:
Got it. Since we started going down this road on this PR, do you think we should also do some checking on line 58 in case the user passes an invalid namespace? Or maybe we prepend the `/` for them in case it's missing?

Reply:
@ivanpauno is exactly right. Without this change, I was seeing runtime errors from some `rcl` / `rclcpp` API calls (though I can't recall exactly which ones).

I guess if a user passes an invalid namespace, they will see similar runtime errors. I think the validation and error message from the underlying layers is enough (it also is nice to keep it encapsulated there in case the naming requirement changes for some reason).

Reply:
Got it, makes sense. I'm just wondering why we haven't seen problems so far with this default value, and whether this change is backwards compatible. But you peeps know best about these details :wink:

Reply:
It's a good question. It looks like the [node initialization logic handles an empty string specifically](https://github.com/ros2/rcl/blob/85ef95514d6ee47bb8285009add84e956595c018/rcl/src/rcl/node.c#L173-L176), but the function for [expanding topic names](https://github.com/ros2/rcl/blob/85ef95514d6ee47bb8285009add84e956595c018/rcl/src/rcl/expand_topic_name.c#L44) does not. It's possible the latter function could be updated in the future to handle this case, but at the end of the day and empty string will resolve to `"/"`.

Reply:
I think this may have messed something up :confused: https://github.com/ros-simulation/gazebo_ros_pkgs/issues/1116

Reply:
Doesn't seem related to me.
That default value is being immediately overridden by the value in the SDF if it exists https://github.com/ros-simulation/gazebo_ros_pkgs/blob/57732fe150efcb324e905f0f68537888c1eceb30/gazebo_ros/src/node.cpp#L57-L59.

Reply:
Maybe, it's related to this https://github.com/ros-simulation/gazebo_ros_pkgs/blob/57732fe150efcb324e905f0f68537888c1eceb30/gazebo_ros/src/gazebo_ros_factory.cpp#L231-L235
https://github.com/ros-simulation/gazebo_ros_pkgs/blob/57732fe150efcb324e905f0f68537888c1eceb30/gazebo_ros/src/gazebo_ros_factory.cpp#L354-L355
?

According to grep that's the only place where we set the `namespace` SDF tag.

Reply:
Nice catch! @ahcorde , are you spawning your robot through the factory? In my case I was.

Reply:
https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1117